File: frontend/src/types/bank-account.ts
import { z } from 'zod';
import { TransactionTypeSchema, PaymentMethodSchema } from './common';

// Bank Account schemas
export const BankAccountSchema = z.object({
  id: z.string().optional(),
  pumpMasterId: z.string(),
  accountHolderName: z
    .string()
    .min(2, 'Account holder name must be at least 2 characters')
    .max(100, 'Account holder name cannot exceed 100 characters'),
  accountNumber: z
    .string()
    .min(10, 'Account number must be at least 10 characters')
    .max(20, 'Account number cannot exceed 20 characters'),
  ifscCode: z.string().length(11, 'IFSC code must be exactly 11 characters'),
  bank: z
    .string()
    .min(2, 'Bank name must be at least 2 characters')
    .max(100, 'Bank name cannot exceed 100 characters'),
  branch: z
    .string()
    .min(2, 'Branch must be at least 2 characters')
    .max(100, 'Branch cannot exceed 100 characters'),
  openingBalance: z.number().min(0, 'Opening balance must be non-negative'),
  openingBalanceDate: z
    .string()
    .regex(/^\d{4}-\d{2}-\d{2}$/, 'Invalid date format'),
  currentBalance: z
    .number()
    .min(0, 'Current balance must be non-negative')
    .nullable()
    .optional(),
  createdAt: z.string().optional(),
  updatedAt: z.string().optional(),
  version: z.number().optional(),
});

export const CreateBankAccountSchema = z.object({
  pumpMasterId: z.string(),
  accountHolderName: z
    .string()
    .min(2, 'Account holder name must be at least 2 characters')
    .max(100, 'Account holder name cannot exceed 100 characters'),
  accountNumber: z
    .string()
    .min(10, 'Account number must be at least 10 characters')
    .max(20, 'Account number cannot exceed 20 characters'),
  ifscCode: z.string().length(11, 'IFSC code must be exactly 11 characters'),
  bank: z
    .string()
    .min(2, 'Bank name must be at least 2 characters')
    .max(100, 'Bank name cannot exceed 100 characters'),
  branch: z
    .string()
    .min(2, 'Branch must be at least 2 characters')
    .max(100, 'Branch cannot exceed 100 characters'),
  openingBalance: z.number().min(0, 'Opening balance must be non-negative'),
  openingBalanceDate: z
    .string()
    .regex(/^\d{4}-\d{2}-\d{2}$/, 'Invalid date format'),
});

export const UpdateBankAccountSchema = CreateBankAccountSchema.partial();

// Bank Transaction schemas
export const BankTransactionSchema = z.object({
  id: z.string().optional(),
  bankAccountId: z.string(),
  amount: z.number().positive('Amount must be positive'),
  transactionType: TransactionTypeSchema,
  paymentMethod: PaymentMethodSchema,
  description: z
    .string()
    .min(1, 'Description is required')
    .max(255, 'Description cannot exceed 255 characters'),
  transactionDate: z.string().optional(),
  entryBy: z.string().optional(),
  createdAt: z.string().optional(),
  updatedAt: z.string().optional(),
});

export const CreateBankTransactionSchema = z.object({
  bankAccountId: z.string(),
  amount: z.number().positive('Amount must be positive'),
  transactionType: TransactionTypeSchema,
  paymentMethod: PaymentMethodSchema,
  description: z
    .string()
    .min(1, 'Description is required')
    .max(255, 'Description cannot exceed 255 characters'),
  transactionDate: z.string().optional(),
  entryBy: z.string().optional(),
});

export const TransactionFormSchema = z.object({
  amount: z.number().positive('Amount must be greater than 0'),
  paymentMethod: PaymentMethodSchema,
  description: z.string().min(1, 'Description is required'),
  transactionDate: z.string().min(1, 'Transaction date is required'),
});

export type BankAccount = z.infer<typeof BankAccountSchema>;
export type CreateBankAccount = z.infer<typeof CreateBankAccountSchema>;
export type UpdateBankAccount = z.infer<typeof UpdateBankAccountSchema>;
export type BankTransaction = z.infer<typeof BankTransactionSchema>;
export type CreateBankTransaction = z.infer<typeof CreateBankTransactionSchema>;
export type TransactionFormValues = z.infer<typeof TransactionFormSchema>;

File: backend/src/main/java/com/reallink/pump/entities/BankAccount.java
package com.reallink.pump.entities;

import java.math.BigDecimal;
import java.time.LocalDate;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.ForeignKey;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import jakarta.persistence.UniqueConstraint;
import jakarta.validation.constraints.DecimalMin;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Size;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Entity
@Table(name = "pump_bank_account_master", uniqueConstraints = {
    @UniqueConstraint(name = "uk_account_number_pump", columnNames = {"account_number", "pump_master_id"}),})
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
public class BankAccount extends BaseEntity {

    @NotNull(message = "Pump master is required")
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "pump_master_id", nullable = false, foreignKey = @ForeignKey(name = "fk_bank_account_pump_master"))
    private PumpInfoMaster pumpMaster;

    @NotBlank(message = "Account holder name is required")
    @Size(min = 2, max = 100, message = "Account holder name must be between 2 and 100 characters")
    @Column(name = "account_holder_name", nullable = false, length = 100)
    private String accountHolderName;

    @NotBlank(message = "Account number is required")
    @Size(min = 10, max = 20, message = "Account number must be between 10 and 20 characters")
    @Column(name = "account_number", nullable = false, length = 20)
    private String accountNumber;

    @NotBlank(message = "IFSC code is required")
    @Size(min = 11, max = 11, message = "IFSC code must be exactly 11 characters")
    @Column(name = "ifsc_code", nullable = false, length = 11)
    private String ifscCode;

    @NotBlank(message = "Bank name is required")
    @Size(min = 2, max = 100, message = "Bank name must be between 2 and 100 characters")
    @Column(name = "bank", nullable = false, length = 100)
    private String bank;

    @NotBlank(message = "Branch is required")
    @Size(min = 2, max = 100, message = "Branch must be between 2 and 100 characters")
    @Column(name = "branch", nullable = false, length = 100)
    private String branch;

    @NotNull(message = "Opening balance is required")
    @DecimalMin(value = "0.00", message = "Opening balance must be greater than or equal to 0.00")
    @Column(name = "opening_balance", nullable = false, precision = 15, scale = 2)
    private BigDecimal openingBalance;

    @NotNull(message = "Opening balance date is required")
    @Column(name = "opening_balance_date", nullable = false)
    private LocalDate openingBalanceDate;
}
